packages/@react-native-windows/cli/src/runWindows/utils/autolink.ts

```diff
@@ -591,6 +592,151 @@ async function updateAutoLink(
+      }
+    }
+  }
+  return changesNeeded;
+}
+
+function fixPackagesConfig(
+  useWinUI3: boolean,
+  project: WindowsProjectConfig,
+  checkMode: boolean,
+) {
+  const projectFile = path.join(
+    project.folder,
+    project.sourceDir,
+    project.project!.projectFile,
+  );
+  const packagesConfig = path.join(
+    path.dirname(projectFile),
+    'packages.config',
+  );
+  console.log(`UseWinUI = ${useWinUI3}`);
+  console.log(`packages.config = ${packagesConfig}`);
+  let changed = false;
+
+  if (fs.existsSync(packagesConfig)) {
+    // if we don't have a packages.config, then this is a C# project, in which case we use <PackageReference> and dynamically pick the right XAML package.
+    const packagesConfigContents = configUtils.readProjectFile(packagesConfig);
+    const packageElements = packagesConfigContents.documentElement.getElementsByTagName(
+      'package',
+    );
+
+    const winuiPropsContents = configUtils.readProjectFile(
+      require.resolve('react-native-windows/PropertySheets/WinUI.props'),
+    );
+    const winui2xVersion = configUtils.tryFindPropertyValue(
+      winuiPropsContents,
+      'WinUI2xVersion',
+    );
+    const winui3Version = configUtils.tryFindPropertyValue(
+      winuiPropsContents,
+      'WinUI3Version',
+    );
+
+    const dialects = [
+      {id: 'Microsoft.WinUI', version: winui3Version!},
+      {id: 'Microsoft.UI.Xaml', version: winui2xVersion!},
+    ];
+    const keepPkg = useWinUI3 ? dialects[0] : dialects[1];
+    const removePkg = useWinUI3 ? dialects[1] : dialects[0];
+
+    const nodesToRemove: Element[] = [];
+    let needsToAddKeepPkg = true;
+    for (let i = 0; i < packageElements.length; i++) {
+      const packageElement = packageElements.item(i)!;
+      const idAttr = packageElement!.getAttributeNode('id');
+      const id = idAttr!.value;
+      if (id === removePkg.id) {
+        nodesToRemove.push(packageElement);
+        packagesConfigContents.documentElement.removeChild(packageElement);
+        changed = true;
+      } else if (id === keepPkg.id) {
+        changed =
+          changed || packageElement.getAttribute('version') !== keepPkg.version;
+        packageElement.setAttribute('version', keepPkg.version!);
+        needsToAddKeepPkg = false;
+      }
+    }
+
+    if (needsToAddKeepPkg) {
+      const newPkg = packagesConfigContents.createElement('package');
+
+      Object.entries(keepPkg).forEach(([attr, value]) => {
+        newPkg.setAttribute(attr, value as string);
+      });
+      newPkg.setAttribute('targetFramework', 'native');
+      packagesConfigContents.documentElement.appendChild(newPkg);
+      changed = true;
+    }
+
+    if (!checkMode && changed) {
+      const serializer = new XMLSerializer();
+      const output = serializer.serializeToString(packagesConfigContents);
+      const formattedXml = formatXml(output);
+      fs.writeFileSync(packagesConfig, formattedXml, {encoding: 'utf-8'});
+    }
+  }
+  return changed;
+}
+
```

Comment:
This file isn't in great shape for this, but it would be great to avoid new 100 line functions.

Reply:
please file a bug to refactor this file